Description: Fix segfault in table_postgres 
 This can occur if the PostgreSQL server suddenly closes the connection 
 at which point PQresultErrorField will return NULL.
Author: Jonas Maurus <debian@maurus.net>
Forwarded: https://github.com/OpenSMTPD/OpenSMTPD-extras/pull/42
--- a/extras/wip/tables/table-postgres/table_postgres.c
+++ b/extras/wip/tables/table-postgres/table_postgres.c
@@ -402,7 +402,9 @@
 
 	if (PQresultStatus(res) != PGRES_TUPLES_OK) {
 		errfld = PQresultErrorField(res, PG_DIAG_SQLSTATE);
-		if (errfld[0] == '0' && errfld[1] == '8') {
+		/* PQresultErrorField can return NULL if the connection to the server
+		   suddenly closed (e.g. server restart) */
+		if (errfld == NULL || (errfld[0] == '0' && errfld[1] == '8')) {
 			log_warnx("warn: table-postgres: trying to reconnect after error: %s",
 			    PQerrorMessage(config->db));
 			PQclear(res);
